### ISSUE TITLE
Moduli Chunking from Better Bootstrapping paper

### ISSUE THREAD

**Issue by jon-chuang** [nucypher/HEAAN.jl#15]
It's probably a good idea to make the early integration of the moduli chunking technique used to speedup keyswitch by more than twice for large parameter sets.

Also, is there going to be GPU support just like for nuFHE? Why is the reference implementation HEAAN not SEAL?

**Comment by fjarri** [comment]
> It's probably a good idea to make the early integration of the moduli chunking technique used to speedup keyswitch by more than twice for large parameter sets.

I agree, and that's something we're aware of - there's already issue #2 and #3 about that. 

> Also, is there going to be GPU support just like for nuFHE? 

Perhaps at some point, but this scheme is not the main priority now.

> Why is the reference implementation HEAAN not SEAL?

To get an initial idea about the scheme it was easier to use the Cheon et al paper and the implementation they wrote. But you have a valid point here, SEAL contains many optimizations that would be a nice addition.

**Comment by jon-chuang** [comment]
I see. What is the intention of HEAAN.jl then? Is it just for prototyping? 

I have an interest in writing a modern HEAAN implementation over a zero-knowledge layer, probably an IOP zk-SNARK (I have yet to decide what the right properties I want yet). Is nuCypher interested in collaborating on this? I believe it would be a world first. This could lead possibly to a research publication and other good things.

**Comment by fjarri** [comment]
We may be interested. Could you write us at ravital@, john@ and bogdan@nucypher.com with a little more info on what you're proposing?
